src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -0,0 +1,217 @@
+    /// Orchestrates test session operations for the engine communicating with the client.
+    /// </summary>
+    public class ProxyTestSessionManager : IProxyTestSessionManager
+    {
+        private readonly object lockObject = new object();
+        private int parallelLevel;
+        private bool skipDefaultAdapters;
+        private Func<ProxyOperationManager> proxyCreator;
+        private Queue<Guid> availableProxyQueue;
+        private IDictionary<Guid, ProxyOperationManagerContainer> proxyMap;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProxyTestSessionManager"/> class.
+        /// </summary>
+        /// 
+        /// <param name="parallelLevel">The parallel level.</param>
+        /// <param name="proxyCreator">The proxy creator.</param>
+        public ProxyTestSessionManager(int parallelLevel, Func<ProxyOperationManager> proxyCreator)
+        {
+            this.parallelLevel = parallelLevel;
+            this.proxyCreator = proxyCreator;
+
+            this.availableProxyQueue = new Queue<Guid>();
+            this.proxyMap = new Dictionary<Guid, ProxyOperationManagerContainer>();
+        }
+
+        /// <inheritdoc/>
+        public void Initialize(bool skipDefaultAdapters)
+        {
+            this.skipDefaultAdapters = skipDefaultAdapters;
+        }
+
+        /// <inheritdoc/>
+        public void StartSession(
+            StartTestSessionCriteria criteria,
+            ITestSessionEventsHandler eventsHandler)
+        {
+            var testSessionInfo = new TestSessionInfo();
+            Task[] taskList = new Task[this.parallelLevel];
+
+            // Create all the proxies in parallel, one task per proxy.
+            for (int i = 0; i < this.parallelLevel; ++i)
+            {
+                taskList[i] = Task.Factory.StartNew(() =>
+                {
+                    // Create the proxy.
+                    var operationManagerProxy = this.CreateProxy();
+
+                    // Initialize the proxy.
+                    operationManagerProxy.Initialize(this.skipDefaultAdapters);
+
+                    // Start the test host associated to the proxy.
+                    operationManagerProxy.SetupChannel(
+                        criteria.Sources,
+                        criteria.RunSettings,
+                        eventsHandler);
+                });
+            }
+
+            // Wait for proxy creation to be over.
+            Task.WaitAll(taskList);
+
+            // Make the session available.
+            TestSessionPool.Instance.AddSession(testSessionInfo, this);
+
+            // Let the caller know the session has been created.
+            eventsHandler.HandleStartTestSessionComplete(testSessionInfo);
+        }
+
+        /// <inheritdoc/>
+        public void StopSession()
+        {
+            foreach (var kvp in this.proxyMap)
+            {
+                // TODO (copoiena): Do nothing for now because in the current implementation the
```

Comment:
If do nothing for now, it is better to comment or remove foreach iteration for unnecessary looping.